tools/serve/serve.py

```diff
@@ -204,6 +207,65 @@ def _script_replacement(self, key, value):
         return None
 
 
+class HtmlScriptInjectorHandlerWrapper:
+    def __init__(self, inject="", wrap=None):
+        self.inject = inject
+        self.wrap = wrap
+
+    def __call__(self, request, response):
+        self.wrap(request, response)
+        # If the response content type isn't html, don't modify it.
+        supported_mime_types = {b"text/html", b"application/xhtml+xml"}
+        if not isinstance(response.headers, ResponseHeaders) or not response.headers.get("Content-Type")[0] in supported_mime_types:
+            return response
+
+        # Skip injection on custom streaming responses.
+        if not isinstance(response.content, (bytes, str, IOBase)) and not hasattr(response, "read"):
+            return response
+
+        # Otherwise, inject the script after the document's doctype.
+        data = "".join(item.decode(response.encoding) for item in response.iter_content(read_file=True))
+
+        # Tokenize and find the position of the first content (e.g. after the
```

Comment:
Can you expand on this comment to describe the rules for where the insertion happens?